### PR TITLE
CH-675/CH-676 added Entities to Order Service

### PR DESCRIPTION
**Pull request by NatalliaBulyha**
jira [create tables and entities + add data] https://cozyhometeam.atlassian.net/browse/CH-676?atlOrigin=eyJpIjoiZmI3NTE1NWQ0OThlNDAxOWIyZmE1NDA4MjFjYWY2OWYiLCJwIjoiaiJ9

### CODE REVIEW

File: src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderDetails.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    private ObjectId id;
+
+    @DBRef
+    private Address address;
+
+    @DBRef
+    private User user;
+
+    @DBRef
+    private OrderItem orderItem;
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderDetails.java]: List<OrderItem> items?

File: src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderDetails.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import org.bson.types.ObjectId;
+import org.springframework.data.mongodb.core.mapping.DBRef;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+@Builder
+@Document(collection = "OrderDetails")
+public class OrderDetails {
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderDetails.java]: maybe will be better to merge order and orderDetails if we use mongodb? in nosql databases normalization is not required

File: src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderDetails.java (hunk 1 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderDetails.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderDetails.java (hunk 2 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderDetails.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/model/AddressDelivery.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+@Getter
+@Setter
+@EqualsAndHashCode(callSuper = true)
+@NoArgsConstructor
+@SuperBuilder
+public class AddressDelivery extends Delivery {
+    private String street;
+    private String house;
+    private Integer apartment;
+    private Integer comment;
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/model/AddressDelivery.java]: Maybe it should be String not Integer?

File: src/main/java/com/cozyhome/onlineshop/orderservice/controller/DeliveryController.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    @GetMapping
+    public ResponseEntity<List<DeliveryCompanyDto>> getDeliveryCompanies() {
+        return new ResponseEntity<>(deliveryService.getDeliveryCompanies(), HttpStatus.OK);
+    }
+
+    @Operation(summary = "Save new delivery company", description = "Save new delivery company")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = SwaggerResponse.Code.CODE_200, description = SwaggerResponse.Message.CODE_200_FOUND_DESCRIPTION) })
+    @PostMapping
+    public ResponseEntity<DeliveryCompanyAdminDto> saveNewDeliveryCompanies(@RequestParam @ValidName String companyName) {
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/controller/DeliveryController.java]: saveNewDeliveryCompany

File: src/main/java/com/cozyhome/onlineshop/orderservice/controller/DeliveryController.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        return new ResponseEntity<>(deliveryService.saveNewDeliveryCompanies(companyName), HttpStatus.OK);
+    }
+
+    @Operation(summary = "Save new delivery company", description = "Save new delivery company")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = SwaggerResponse.Code.CODE_200, description = SwaggerResponse.Message.CODE_200_FOUND_DESCRIPTION) })
+    @DeleteMapping
+    public ResponseEntity<DeliveryCompanyAdminDto> deleteDeliveryCompanies(@RequestParam @ValidName String companyName) {
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/controller/DeliveryController.java]: deleteDeliveryCompany

File: src/main/java/com/cozyhome/onlineshop/orderservice/controller/OrderController.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    private final OrderService orderService;
+    @Value("${header.name.user-id}")
+    private String userIdAttribute;
+
+    @Operation(summary = "Create new order", description = "Create new order")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = SwaggerResponse.Code.CODE_200, description = SwaggerResponse.Message.CODE_200_FOUND_DESCRIPTION) })
+    @PostMapping
+    public ResponseEntity<OrderNumberDto> newOrder(@RequestBody OrderDto orderDto, HttpServletRequest request) {
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/controller/OrderController.java]: "createOrder" or "saveOrder", the method should be a verb

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/DeliveryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    private final ModelMapper modelMapper;
+    @Value("${status}")
+    private EntityStatus status;
+
+    @Override
+    public List<DeliveryCompanyDto> getDeliveryCompanies() {
+        List<DeliveryCompany> deliveryCompanies = deliveryCompanyRepository.findAllByStatus(status);
+        if (deliveryCompanies.isEmpty()) {
+            return new ArrayList<>();
+        }
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/DeliveryServiceImpl.java:33]: what need to return empty collection if collection is empty? :)

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        List<OrderItem> orderItems = orderItemBuilder.buildOrderItems(orderDto.getOrderItems());
+
+        Order order = Order.builder()
+                .firstName(orderDto.getFirstName())
+                .lastName(orderDto.getLastName())
+                .phoneNumber(orderDto.getPhoneNumber())
+                .orderItem(orderItems)
+                .orderStatus(OrderStatus.NEW)
+                .createdAt(LocalDateTime.now())
+                .modifiedAt(LocalDateTime.now())
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/OrderServiceImpl.java:43]: should we fill this field when create a new order?

File: src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderNumber.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.time.LocalDate;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+@Builder
+@Document(collection = "OrderNumber")
+public class OrderNumber {
+    private ObjectId id;
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/model/OrderNumber.java]: can we replace OrderNumber to order?

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                .build();
+
+        if (orderDto.getDelivery().getStreet() == null) {
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/OrderServiceImpl.java]: more obvious to check if (deliveryCompanyName != null)

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/DeliveryBuilder.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    public Delivery buildPostalDelivery(DeliveryDto deliveryDto) {
+        DeliveryCompany deliveryCompany = deliveryCompanyRepository.findByNameAndStatus(deliveryDto.getDeliveryCompanyName(),
+                EntityStatus.ACTIVE)
+                .orElseThrow(() -> new DataNotFoundException(String.format("Delivery company %s doesn't found", deliveryDto.getDeliveryCompanyName())));
+
+        Delivery delivery = PostalDelivery.builder()
+                .city(deliveryDto.getCity())
+                .paymentMethod(deliveryDto.getPaymentMethod())
+                .createdAt(LocalDateTime.now())
+                .modifiedAt(LocalDateTime.now())
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/DeliveryBuilder.java]: should we fill this field during creating a delivery?

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/DeliveryBuilder.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+        return deliveryRepository.save(delivery);
+    }
+
+    public Delivery buildAddressDelivery(DeliveryDto deliveryDto) {
+        Delivery delivery = AddressDelivery.builder()
+                .city(deliveryDto.getCity())
+                .paymentMethod(deliveryDto.getPaymentMethod())
+                .createdAt(LocalDateTime.now())
+                .modifiedAt(LocalDateTime.now())
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/DeliveryBuilder.java]: should we fill this field during creating a delivery?

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/OrderItemBuilder.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    private OrderItem buildOrderItem(OrderItemDto orderItemDto) {
+        Inventory inventory = inventoryRepository.findByProductColorProductSkuCodeAndProductColorColorHex(
+                orderItemDto.getProductSkuCode(), orderItemDto.getColorHex())
+                .orElseThrow(() -> new DataNotFoundException(String.format("Product with sku code %s and color %s " +
+                        "doesn't found", orderItemDto.getProductSkuCode(), orderItemDto.getColorHex())));
+
+        if (inventory.getQuantity() < orderItemDto.getQuantity()) {
+         throw new ProductOutOfStockException("Product with sku code %s and color hex %s is out of stock.");
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/OrderItemBuilder.java]: it also can be "Not enough quantity to order", out of stock is when a quantity is 0.

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/OrderItemBuilder.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        }
+        inventory.setQuantity(inventory.getQuantity() - orderItemDto.getQuantity());
+        //inventory.setQuantity(2000);
+        inventoryRepository.save(inventory);
+        OrderItem orderItem = OrderItem.builder()
+                .productColorId(inventory.getId())
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/OrderItemBuilder.java]: maybe "inventory.getProductColor().getId()"?

File: src/main/java/com/cozyhome/onlineshop/orderservice/model/AddressDelivery.java (hunk 3 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/model/AddressDelivery.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/controller/DeliveryController.java (hunk 4 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/controller/DeliveryController.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/controller/DeliveryController.java (hunk 5 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/controller/DeliveryController.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/controller/OrderController.java (hunk 6 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/controller/OrderController.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/DeliveryServiceImpl.java (hunk 7 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/DeliveryServiceImpl.java:33]: мы не можем отправить deliveryCompanies, потому что контроллер ждет DeliveryCompanyDto, нужно мапить, на мой взгляд return new ArrayList<>() более емко говорит что мы отправляем пустой список

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/OrderServiceImpl.java (hunk 10 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/service/impl/OrderServiceImpl.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/OrderItemBuilder.java (hunk 13 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/OrderItemBuilder.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/OrderItemBuilder.java (hunk 14 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/OrderItemBuilder.java]: done

File: src/main/java/com/cozyhome/onlineshop/userservice/security/JWT/JwtTokenFilter.java [hunk 15]
```diff
@@ -23,45 +23,43 @@
...
+    private JwtTokenUtil jwtTokenUtil;
+    @Autowired
+    private ExtendedUserDetailsService userDetailsService;
     @Value("${header.name.user-id}")
     private String userIdAttributeName;
     @Value("${header.name.user-role}")
     private String userRoleAttributeName;
-    @Value("${api.basePath}")
+    @Value("${api.secure.basePath}")
     private String noAuthPathUrl;
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/userservice/security/JWT/JwtTokenFilter.java]: authPathUrl

File: src/main/java/com/cozyhome/onlineshop/userservice/security/JWT/JwtTokenFilter.java [hunk 16]
```diff
@@ -23,45 +23,43 @@
...
-                log.info("[ON doFilterInternal]:: jwtToken [{}]", jwtToken);
-                if (jwtTokenUtil.isTokenInBlackList(jwtToken)) {
-                    log.warn("[ON doFilterInternal]:: Token is in Black List. Access denied");
-                    throw new AuthException("User logged out. Access denied.");
-                }
-                username = jwtTokenUtil.getUsernameFromToken(jwtToken);
-                log.info("[ON doFilterInternal]:: username [ {} ]", username);
+        String jwtToken = jwtTokenUtil.resolveToken(request);
+        String username = null;
+        if (shouldFilter(request) || jwtToken != null) {
```
Comment (DariaBoo) [inline src/main/java/com/cozyhome/onlineshop/userservice/security/JWT/JwtTokenFilter.java:42]: good idea!)

File: src/main/java/com/cozyhome/onlineshop/userservice/security/JWT/JwtTokenFilter.java (hunk 15 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/userservice/security/JWT/JwtTokenFilter.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/DeliveryBuilder.java (hunk 11 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/DeliveryBuilder.java]: done

File: src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/DeliveryBuilder.java (hunk 12 above)
Comment (NatalliaBulyha) [reply src/main/java/com/cozyhome/onlineshop/orderservice/service/builder/DeliveryBuilder.java]: done

**Pull request merged**